### PR TITLE
[GlobalIsel][AArch64] fix out of range access in regbankselect

### LINKED ISSUE #92062: [RegBankSelect] Assertion `N < NumContainedTys && "Element number out of range!"'
I tried the following code:

```llvm
target triple = "arm64"

@p = external global { {}, { ptr } }

define void @foo() {
bb:
  %i1 = load ptr, ptr @p, align 8
  store ptr %i1, ptr null, align 8
  ret void
}
```

When using llc in LLVM 18, the following error occurs:

```
llc: /root/llvm-project/llvm/include/llvm/IR/DerivedTypes.h:343: llvm::Type* llvm::StructType::getElementType(unsigned int) const: Assertion `N < NumContainedTys && "Element number out of range!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/llc -o /app/output.s -x86-asm-syntax=intel -O0 <source>
1.	Running pass 'Function Pass Manager' on module '<source>'.
2.	Running pass 'RegBankSelect' on function '@foo'
...
 #8 0x0000000000b27667 llvm::AArch64RegisterBankInfo::isLoadFromFPType(llvm::MachineInstr const&) const (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0xb27667)
 #9 0x0000000000b2b155 llvm::AArch64RegisterBankInfo::getInstrMapping(llvm::MachineInstr const&) const (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0xb2b155)
#10 0x0000000004203f47 llvm::RegBankSelect::assignInstr(llvm::MachineInstr&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x4203f47)
#11 0x00000000042044d3 llvm::RegBankSelect::assignRegisterBanks(llvm::MachineFunction&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x42044d3)
...
Compiler returned: 139
```

https://llvm.godbolt.org/z/a4h1abfKc

### PR DESCRIPTION
**Pull request by tschuett**
Fixes https://github.com/llvm/llvm-project/issues/92062

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/pr92062.ll [hunk 1]
```diff
@@ -0,0 +1,21 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=aarch64 -O0 -global-isel -global-isel-abort=2 -verify-machineinstrs %s -o - 2>&1 | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/pr92062.ll:2]:
```suggestion
; RUN: llc -mtriple=aarch64 -O0 -global-isel %s -o - 2>&1 | FileCheck %s
```

Do you really need the -O0?

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/AArch64/pr92062.ll (hunk 1 above)
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/pr92062.ll:2]: Unfortunately, I need the -O0.

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp [hunk 2]
```diff
@@ -600,8 +600,11 @@ bool AArch64RegisterBankInfo::isLoadFromFPType(const MachineInstr &MI) const {
     EltTy = GV->getValueType();
     // Look at the first element of the struct to determine the type we are
```
Comment (DianQK) [inline llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:601]:
```suggestion
    // Look at the first valid element of the struct to determine the type we are
```

It seems we should also update the comment.

**Review by dianqk — APPROVED** [review]
Thank you for the quick fix.

**Pull request merged**